### ISSUE TITLE
Placement references to clusters violates API convention

### ISSUE THREAD

**Issue by pmorie** [kubernetes-sigs/federation-v2#688]
TLDR: Cluster names for the purposes of placement are specified unqualified, but according to kube api conventions that means a given cluster should reside in the same namespace as the federated resource that refers to it.

### Background Info

In the history of federation-v2, the `Cluster` resource in the cluster-registry and `FederatedCluster` resource in federation-v2 were originally cluster-scoped. This was reflective of the scoping of the `Cluster` resource in the original federation effort. Another historical factor is the identification of clusters in federation APIs by name only, which was entirely sufficient due to their cluster-scoped nature. Finally, originally federation-v2 was conceived of as running exclusively in a cluster-scoped mode where a single controller serviced all the namespaces in a cluster. 

Later in time, three things changed:

- The `Cluster` resource was changed to be namespace-scoped
- Correspondingly, the `FederatedCluster` resource was changed to be namespace-scoped
- We introduced the ability to deploy the federation controllers scoped to a single namespace

These changes introduced a wrinkle from an API perspective: the semantics of identifying clusters in federated APIs depend on how federation is deployed:

- When federation is deployed scoped to a namespace, `spec.placement.clusterNames` is a list of the names of `FederatedClusters` in the containing namespace
- When federation is deployed scoped to the entire cluster, `spec.placement.clusterNames` is a list of the names of `FederatedClusters` in a namespace whose value is a configuration item of the federation controller.

According to the [API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#object-references) guidance on object references, object references made using just the name should correspond to objects within the containing namespace. We adhere to this guidance when federation is deployed scoped to a namespace; we don't adhere when federation is deployed in the cluster scope.

A simple thing that we can do to adhere to this guidance for ns-scoped types is for name-only references always to refer to things within the containing namespace. While this would require changes to the controller, it would not affect the existing API surface for ns-scoped resources.

### Example

Let's consider the example of a `FederatedDeployment` in the `example` namespace:

```yaml
apiVersion: types.federation.k8s.io/v1alpha1
kind: FederatedDeployment
metadata:
  name: hello
  namespace: example
spec:
  placement:
    clusterNames:
    - cluster-x
    - cluster-y
  # rest omitted
```

The `FederatedClusters` addressed by `spec.placement.clusterNames` are ambiguous; it might be:

- `cluster-x` and `cluster-y` in the `example` namespace
- `cluster-x` and `cluster-y` in the configured federation namespace of the controller

**Comment by xunpan** [comment]
There has a similar issue for `Cluster` and `FederatedCluster` references. Now, we plan to make both resources to the same namespace by default https://github.com/kubernetes-sigs/federation-v2/issues/114#issuecomment-475130017.  

However, it still has case in sharing `Cluster` or `FederatedCluster` from different namespaces. 

Keeping current behavior seems not good for usability because convention break. 
To follow up convention, there are 3 ways in my mind:
 - namespace/reference
    duplicated informaiton need to be specified.
 - global resources for sharing in different 
    a little heavy for obvious sharing a resource from federation deployment
 - duplicate clusters in namespaces
    bad
I think all of them are not good enough.

Any comments @marun, @irfanurrehman , @shashidharatd

**Comment by pmorie** [comment]
@xunpan Just want to make sure I understand your comments correctly


> - namespace/reference

This refers to allowing explict cross-ns references, correct? (ie, ability to specify explicitly the namespace of a referred cluster when you want to reference a FederatedCluster outside the enclosing namespace)

> - global resources for sharing in different

Not sure what you meant here - can you elaborate?

> - duplicate clusters in namespaces

I think this refers to, allow name-only references only within the enclosing namespace, and the 'duplicate' comes from having to create a FederatedCluster in each namespace you want to spread resources to that cluster from, is that accurate?

**Comment by pmorie** [comment]
In the April 24 2019 meeting, we established consensus that explicit name-based references should be conformant with the API conventions. Discussion still happening on selectors.

**Comment by marun** [comment]
A review of this issue with @deads2k, @pmorie and myself concluded that explicit namespace references are not required, but that the use of a string representing a cluster name to reference a cluster in placement directives should be changed to an object with a name field.

Here is the current method of referencing clusters in placement:

```yaml
spec: 
  placement:
    clusterNames:
      - cluster1
      - cluster2
```

This needs to become:

```yaml
spec:
  placement:
    clusters:
      - name: cluster1
      - name: cluster2
```

This change will have minimal user impact, is consistent with how clusters are referenced in propagation status (#843), and leaves the door open to adding additional fields (e.g. UID, namespace) to a reference if necessary in the future.

**Comment by marun** [comment]
I'm going to take on implementing this change.

**Comment by font** [comment]
What about the use of `clusterName` in `ClusterObjectVersion`? It seems we should also use a `Cluster` object.

https://github.com/kubernetes-sigs/federation-v2/blob/2d354db6e977571bfa374881293a08955eeb9c68/pkg/apis/core/v1alpha1/propagatedversion_types.go#L37-L43

**Comment by font** [comment]
Similarly for `FederatedServiceClusterStatus`:
https://github.com/kubernetes-sigs/federation-v2/blob/2d354db6e977571bfa374881293a08955eeb9c68/pkg/apis/core/v1alpha1/federatedservicestatus_types.go#L25-L28

**Comment by marun** [comment]
@font I think those a `clusterName` string needs to be converted to a `Cluster` object in those cases.  A slice of strings was a problem because there's no easy way to add new fields.  The examples you cite are already objects, so adding new fields is not a problem.  If anything it might be worth renaming the field from `ClusterName` to `Name`, since the context should make it obvious what it is a name of.
